airmail_indexer/src/query_pip.rs

```diff
@@ -215,92 +127,74 @@ async fn query_names(url: &str) -> Option<Vec<String>> {
         .iter()
         .cloned()
         .collect::<Vec<_>>();
-    Some(names)
+
+    Some((admin_id, names))
 }
 
-async fn query_langs(url: &str) -> Option<Vec<String>> {
-    let response = if let Ok(response) = HTTP_CLIENT
-        .with(|client: &reqwest::Client| client.get(url).send())
-        .await
-    {
-        response
-    } else {
-        return None;
-    };
-    if response.status() != 200 {
-        return None;
-    }
-    let response = if let Ok(response) = response.text().await {
-        response
-    } else {
-        return None;
-    };
-    let response: PipPropertyResponse = if let Ok(response) = serde_json::from_str(&response) {
-        response
-    } else {
-        return None;
-    };
-    response
+async fn query_langs(country_id: u64, wof_db: &WhosOnFirst) -> Option<(u64, Vec<String>)> {
+    let response: PipLangsResponse = wof_db.properties_for_id(country_id).await.ok()?.into();
+    let langs: Vec<String> = response
         .langs
-        .map(|langs| langs.split(',').map(|s| s.to_string()).collect())
+        .map(|langs| langs.split(',').map(|s| s.to_string()).collect())?;
+
+    Some((country_id, langs))
 }
 
 pub(crate) async fn query_pip(
-    read: &'_ ReadTransaction<'_>,
+    indexer_cache: &IndexerCache,
     to_cache_sender: Sender<WofCacheItem>,
     s2cell: u64,
-    port: usize,
-) -> Result<PipResponse, Box<dyn Error>> {
-    let wof_ids = query_pip_inner(s2cell, read, to_cache_sender.clone(), port).await?;
-    let mut name_handles: Vec<(u64, JoinHandle<Option<Vec<String>>>)> = Vec::new();
-    let mut lang_handles: Vec<(u64, JoinHandle<Option<Vec<String>>>)> = Vec::new();
-    let mut cached_names = Vec::new();
-    let mut cached_langs = Vec::new();
+    wof_db: &WhosOnFirst,
+    pip_tree: &Option<PipTree<ConcisePipResponse>>,
+) -> Result<PipResponse> {
+    let wof_ids = query_pip_inner(
+        s2cell,
+        indexer_cache,
+        to_cache_sender.clone(),
+        wof_db,
+        pip_tree,
+    )
+    .await?;
+    let mut response = PipResponse::default();
+    let mut admin_name_futures = vec![];
+    let mut lang_futures = vec![];
+
+    // Query names for the admin areas
     for admin_id in wof_ids.all_admin_ids {
-        if let Ok(names) = query_names_cache(read, &admin_id) {
-            cached_names.extend(names);
-        } else {
-            let names_url = format!("http://localhost:{}/place/wof/{}/name", port, &admin_id);
-            let handle: JoinHandle<Option<Vec<String>>> =
-                tokio::spawn(async move { query_names(&names_url).await });
-            name_handles.push((admin_id, handle));
-        }
+        // This check was at the end, but I think it should be here as the ID has already been looked up
```

Comment:
nit: justification makes sense, no need to reference the way the code was before.

Reply:
Thanks - yes, glad you picked this up. I wasn't sure if there was something happening after this step which required pushing the IDs in, but not processing further. I didn't have the confidence that there wasn't something happening after this step which required pushing the IDs in still. I'll remove